### ISSUE TITLE
fetch all pages until end for `get_storage_keys_paged`

### ISSUE THREAD

**Issue by brenzi** [scs/substrate-api-client#588]
instead of asking for a max_count, max count could be assumed to be 1000 (seems to be rpc default) and then just fetch all pages directly

**Comment by haerdib** [comment]
I think a max_key input should still be accepted in case someone wants to get a limited number of keys, like for performance reason (?). But we could provide a second function that simply fetches all keys with the given prefix.

Could look something like that:
```rust
async fn get_storage_keys_paged_all(
		&self,
		prefix: Option<StorageKey>,
		start_key: Option<StorageKey>,
		at_block: Option<Self::Hash>,
	) -> Result<Vec<StorageKey>>{ 
	let max_keys = 1000;
        let storage_keys = Vec::new();
        for i in u32::Max {
	       let keys = self.get_storage_keys_paged(Some(prefix), max_keys, start_key, at_block)?;
               storage_keys.push(keys);
               // No more keys available for the given prefix, returning.
	       if keys.len() < max_keys as usize {
    	              return Ok(storage_keys);
	       }
	}
}



	
```

**Comment by haerdib** [comment]
After thinking about it some, contrary to the above, I believe the default version should be failsafe. But let's keep the count as input value. That way the user still has the full flexibility, can choose the max keys to fetch but without the (runtime) limit of 1000. The function will internally split the key fetch into slices of 1000. The drawback: performance, in case one could fetch more than 1000 keys. For these (probably rare usecases) we could still offer a "bare" function, but with a different name.

Idea:
```rust
async fn get_storage_keys_paged(
		&self,
		prefix: Option<StorageKey>,
                count: u32,
		start_key: Option<StorageKey>,
		at_block: Option<Self::Hash>,
	) -> Result<Vec<StorageKey>>{ 
        // Substrate node does normally only allow 1000 keys as max value.
	let max_keys_per_fetch = 1000;
        let storage_keys = Vec::new();
        for slice_of_1000_or_below until `count` is reached {
        	let keys = self.get_storage_keys_paged_bare(prefix, slice_of_1000_or_below, start_key, at_block)?;
                storage_keys.push(keys);
	}
       return Ok(storage_keys);
}
```

where `get_storage_keys_paged_bare` is the original, renamed function.

**Comment by echevrier** [comment]
We can't fetch all pages until end. Substrate does not to support this feature. We can only fetch a certain number of keys. See Discussion in PR https://github.com/paritytech/substrate/issues/4717. If we want achieve this, we have to create an issue in Substrate.

Solution proposed by @haerdib allows to fetch a certain number greater than 1000 of keys. As `count` is mandatory for Substrate, either the user chooses the max number of keys to fetch, or we defined this value. I prefer to leave this choice to the user. But are we sure that there is no limit? What is happening for a very large number? Ex 10'000?

@haerdib what do you mean with `The drawback: performance, in case one could fetch more than 1000 keys. For these (probably rare usecases) we could still offer a "bare" function, but with a different name.` ? Furthermore, is your `bare `function not the current one, which throws an error for `count `bigger than 1000?

**Comment by haerdib** [comment]
We can offer this without Substrate support by building a wrapper function around the rpc call.

So, yes, exactly. The `bare` function is the current one. My proposal is to basically build a Wrapper function around it. E.g. when the user chooses `10'000` as max number (and there as that many keys) the `bare` function would get called 10 times. Each time with the start key+1000 compared to the previous call.

Does that make sense?

**Comment by haerdib** [comment]
Regarding the drawback: In case Substrate updates the accepted default, or if the users itself updates it (if that's possible), our wrapper function would still split the call into multiple calls if the number is larger than 1000. That might not be ideal for the user. So that's why we could keep the `bare` function, where the user is responsible to check that the #count values can be fetched from Substrate.

**Comment by echevrier** [comment]
> We can offer this without Substrate support by building a wrapper function around the rpc call. 

Not exactly. What we can offer is that the user can fetch a certain number greater than 1000 of keys. Not all pages until end. To get this, the user has to know how many keys are in the storage.

> So, yes, exactly. The `bare` function is the current one. My proposal is to basically build a Wrapper function around it. E.g. when the user chooses `10'000` as max number (and there as that many keys on substrate side) the `bare` function would get called 10 times. Each time with the start key+1000 compared to the previous call.
> Does that make sense?
Yes, if we don't have performance problem. I implemented something like that. See https://github.com/scs/substrate-api-client/pull/602. I haven't change `async fn get_storage_keys_paged`. This is the `bare` function and add `get_all_storage_keys_paged_up_to_count`

But are we sure that no limit is necessary?What happens if there is a very large number of calls? 1000 calls?

**Comment by haerdib** [comment]
Cool, thanks a lot! I'll comment in the PR then :)

> But are we sure that no limit is necessary? What happens if there is a very large number of calls? 1000 calls?
Then 1000 calls will be made. That's in the responsibility of the user, I believe. However, keep in mind that `get_storage_keys_paged` only returns the keys that start with the given prefix. So even if you use `10000` (rsp. u32::max) as `count` input, if there's less key with given prefix in the storage, only those keys are returned, not the given max count.
